Updated to work with Expressive (future) 2.0

weierophinney opened the pull request:
Updates to allow working with Stratigility 2.0 via the Expressive develop branch (future 2.0). This means that middleware should be http-interop/http-middleware types as much as possible, and also means that default invocation is via `process()`.

This is a work in progress until that feature branch is merged to Expressive's develop branch and/or tagged as version 2.0.

samsonasik commented:
Just tried it, in PHP 7.1 and got the following error:

```web
Fatal error: Uncaught Error: Call to a member function process() on string 
in /vendor/zendframework/zend-expressive/src/Middleware/DispatchMiddleware.php on line 43
```

weierophinney commented:
@samsonasik Can you do the following for me, please?

- How are you piping and/or routing middleware? Can you provide examples?
- Related, can you provide the dependency (container) configuration for the piped and routed middleware?
- What is the output of `composer show`?

samsonasik commented:
@weierophinney here is what I tried:
- I use default skeleton (zf-servicemanager, zend-router, zend-view), the HomeAction same as https://github.com/weierophinney/zend-expressive-skeleton/blob/067f39f785579a5ff39fdf0deed63f20174a7ef6/src/App/src/Action/HomePageAction.php
- I used default skeleton for dependency container config (nothing changed)
- Here is my composer.json config:
```json
    "require": {
        "php": "^5.6 || ^7.0",
        "roave/security-advisories": "dev-master",
        "zendframework/zend-expressive": "dev-feature/stratigility-2.0 as 1.1.0",
        "zendframework/zend-expressive-helpers": "^3.0",
        "zendframework/zend-stdlib": "^2.7 || ^3.0",
        "zendframework/zend-expressive-zendrouter": "^2.0",
        "zendframework/zend-servicemanager": "^2.7.3 || ^3.0",
        "zendframework/zend-expressive-zendviewrenderer": "^1.0"
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/weierophinney/zend-expressive.git"
        }
    ],
```
- Here is composer show output:
```bash
$ composer show
container-interop/container-interop            1.1.0                                Promoting the interoperability of container objects (DIC, SL, etc.)
doctrine/instantiator                          1.0.5                                A small, lightweight utility to instantiate objects in PHP without invoking their constructors
fig/http-message-util                          1.1.0                                Utility classes and constants for use with PSR-7 (psr/http-message)
filp/whoops                                    2.1.5                                php error handling for cool kids
http-interop/http-middleware                   0.4.1                                Common interface for HTTP server-side middleware
phpdocumentor/reflection-common                1.0                                  Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock              3.1.1                                With this component, a library can provide support for annotations via DocBlocks or otherwise retrieve information that is...
phpdocumentor/type-resolver                    0.2.1                                
phpspec/prophecy                               v1.6.2                               Highly opinionated mocking framework for PHP 5.3+
phpunit/php-code-coverage                      2.2.4                                Library that provides collection, processing, and rendering functionality for PHP code coverage information.
phpunit/php-file-iterator                      1.4.2                                FilterIterator implementation that filters files based on a list of suffixes.
phpunit/php-text-template                      1.2.1                                Simple template engine.
phpunit/php-timer                              1.0.8                                Utility class for timing
phpunit/php-token-stream                       1.4.9                                Wrapper around PHP's tokenizer extension.
phpunit/phpunit                                4.8.34                               The PHP Unit Testing framework.
phpunit/phpunit-mock-objects                   2.3.8                                Mock Object library for PHPUnit
psr/http-message                               1.0.1                                Common interface for HTTP messages
roave/security-advisories                      dev-master 3db4b0d                   Prevents installation of composer packages with known security vulnerabilities: no API, simply require it
sebastian/comparator                           1.2.4                                Provides the functionality to compare PHP values for equality
sebastian/diff                                 1.4.1                                Diff implementation
sebastian/environment                          1.3.8                                Provides functionality to handle HHVM/PHP environments
sebastian/exporter                             1.2.2                                Provides the functionality to export PHP variables for visualization
sebastian/global-state                         1.1.1                                Snapshotting of global state
sebastian/recursion-context                    1.0.2                                Provides functionality to recursively process PHP variables
sebastian/version                              1.0.6                                Library that helps with managing the version number of Git-hosted PHP projects
squizlabs/php_codesniffer                      2.7.1                                PHP_CodeSniffer tokenizes PHP, JavaScript and CSS files and detects violations of a defined set of coding standards.
symfony/yaml                                   v3.2.2                               Symfony Yaml Component
webmozart/assert                               1.2.0                                Assertions to validate method input/output with nice error messages.
zendframework/zend-diactoros                   1.3.10                               PSR HTTP Message implementations
zendframework/zend-escaper                     2.5.2                                
zendframework/zend-eventmanager                3.1.0                                Trigger and listen to events within a PHP application
zendframework/zend-expressive                  dev-feature/stratigility-2.0 580351c PSR-7 Middleware Microframework based on Stratigility
zendframework/zend-expressive-helpers          3.0.1                                Helper/Utility classes for Expressive
zendframework/zend-expressive-router           2.1.0                                Router subcomponent for Expressive
zendframework/zend-expressive-template         1.0.4                                Template subcomponent for Expressive
zendframework/zend-expressive-zendrouter       2.0.0                                zend-mvc router support for Expressive
zendframework/zend-expressive-zendviewrenderer 1.2.1                                zend-view PhpRenderer integration for Expressive
zendframework/zend-http                        2.6.0                                provides an easy interface for performing Hyper-Text Transfer Protocol (HTTP) requests
zendframework/zend-loader                      2.5.1                                
zendframework/zend-psr7bridge                  0.2.2                                PSR-7 <-> Zend\Http bridge
zendframework/zend-router                      3.0.2                                
zendframework/zend-servicemanager              3.2.0                                
zendframework/zend-stdlib                      3.1.0                                
zendframework/zend-stratigility                2.0.1                                Middleware for PHP
zendframework/zend-uri                         2.5.2                                a component that aids in manipulating and validating » Uniform Resource Identifiers (URIs)
zendframework/zend-validator                   2.8.2                                provides a set of commonly needed validators
zendframework/zend-view                        2.8.1                                provides a system of helpers, output filters, and variable escaping
```

weierophinney commented:
@samsonasik I've just done the following:

- First, I created a `packages.json` with the following contents:

  ```json
  {
      "packages": {
          "zendframework/zend-expressive-skeleton": {
              "1.1.0": {
                  "name": "zendframework/zend-expressive-skeleton",
                  "version": "1.1.0",
                  "source": {
                      "type": "git",
                      "url": "https://github.com/weierophinney/zend-expressive-skeleton.git",
                      "reference": "feature/stratigility-2.0"
                  }
              }
          }
      }
  }
  ```

- Next, I invoked the following:

  ```bash
  $ composer create-project --repository packages.json zendframework/zend-expressive-skeleton zend-expressive-skeleton-129 1.1.0
  ```

- During installation, I told it to:
  - Use the "Flat" installation (default)
  - Use zend-servicemanager (default)
  - Use zend-router (non-default, but what you appear to have selected)
  - Use zend-view (non-default; again, your selection)
  - Use whoops (you did not indicate your preference for errors)
  - Install zend-validator to `config/config.php`, as well as any other packages of that type

- Ran `composer serve` from the newly created directory.

I did all of these commands using PHP 7.1.0.

I then visited http://localhost:8080, and the home page came up fine, as did `/api/ping`.

I noticed a few differences between my installation and yours, but only around PHPUnit and its dependencies; all others were the same.

As such, I cannot reproduce at this time.

xtreamwayz commented:
@samsonasik I have trying to reproduce it but it's working over here. Are you sure you created a propject from the installer located at: https://github.com/weierophinney/zend-expressive-skeleton/tree/feature/stratigility-2.0?

To test it I locally checkout that branch from weierophinney/zend-expressive-skeleton and after that I use this: https://gist.github.com/xtreamwayz/7590e55f95e98daf8d930b64cd8e4b69

xtreamwayz commented on src/App/src/Action/HomePageAction.php:
```diff
@@ -2,6 +2,8 @@
 
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use Interop\Http\ServerMiddleware\MiddlewareInterface as ServerMiddlewareInterface;
 use Psr\Http\Message\ResponseInterface;
```
This line is obsolete: `ResponseInterface` is not used.

xtreamwayz commented on src/App/src/Action/PingAction.php:
```diff
@@ -2,13 +2,15 @@
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use Interop\Http\ServerMiddleware\MiddlewareInterface as ServerMiddlewareInterface;
 use Zend\Diactoros\Response\JsonResponse;
 use Psr\Http\Message\ResponseInterface;
```
This line is obsolete: ResponseInterface is not used.

xtreamwayz commented on test/AppTest/Action/HomePageActionTest.php:
```diff
@@ -3,11 +3,17 @@
-use Zend\Diactoros\Response;
-use Zend\Diactoros\ServerRequest;
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use PHPUnit\Framework\Assert;
```
`PHPUnit\Framework\Assert` is not used in this test.

weierophinney commented:
@xtreamwayz Thanks for the review; incorporated all feedback at this time.

samsonasik commented:
@weierophinney thank you. I confirm that it works. It seems I was using default `composer create-project` which read form dev-master.

xtreamwayz commented:
Just thinking out loud:

Can we change in `test/ExpressiveInstallerTest/ProjectSandboxTrait.php`

```php
        // Import programmatic/declarative middleware pipeline and routing configuration statements
        $app->pipe(ErrorHandler::class);
        $app->pipe(ServerUrlMiddleware::class);
        $app->pipeRoutingMiddleware();
        $app->pipe(ImplicitHeadMiddleware::class);
        $app->pipe(ImplicitOptionsMiddleware::class);
        $app->pipe(UrlHelperMiddleware::class);
        $app->pipeDispatchMiddleware();
        $app->pipe(NotFoundHandler::class);

        if ($setupRoutes === true && $container->has(HomePageAction::class)) {
            $app->get('/', HomePageAction::class, 'home');
        }

        if ($setupRoutes === true && $container->has(PingAction::class)) {
            $app->get('/api/ping', PingAction::class, 'api.ping');
        }
```

with

```php
        // Import programmatic/declarative middleware pipeline and routing configuration statements
        require 'config/pipeline.php';
        if (file_exists('config/routes.php')) {
            require 'config/routes.php';
        }
```

Also the signature can be changed from `getAppResponse($path = '/', $setupRoutes = true)` to `getAppResponse($path = '/')`. It will setup routes if the file exists and if it actually contains any routes.

It seems to work in a local project and it makes sure all the latest changes to piped middleware and routes config files are tested.

RalfEggert commented:
@weierophinney 

I used your steps to install the new skeleton from this comment:
https://github.com/zendframework/zend-expressive-skeleton/pull/129#issuecomment-276458486

I discussed the new structure of the skeleton with some guys who want to start with `Zend\Expressive` and are waiting for the new modular structure. The flat structure is no option for them. Here are some ideas for improvement from yesterdays discussion:

- Why is the `/src/` directory called `src` at all? It contains classes and templates and the configuration for this module. We expected a `/module/` or `/modules/` directory instead.

- Why is some of the configuration for the `App` module still placed in `/config/autoload/`? For example, the `dependencies` configuration for the two action middleware `PingAction` and `HomePageAction` is still placed in `/config/autoload/routes.global.php` and should be deleted since it is already defined in the `App/ConfigProvider`. It is also confusing that some of the layout config is in the `/config/autoload/templates.global.php` while the rest is in the `App/ConfigProvider` class. Please put it in one place only.

- If the `dependencies` configuration for the two action middleware `PingAction` and `HomePageAction`  is removed from the `/config/autoload/routes.global.php` file then the `RouterInterface::class` configuration could be moved to the `/config/autoload/dependencies.global.php` file and the `/config/autoload/routes.global.php` file could be deleted altogether, couldn't it? Same for the `/config/autoload/templates.global.php` when the `templates` config is moved to the `App\ConfigProvider` class. Less config file makes it easier for beginners to adopt `Zend\Expressive`.

- I find it really confusing that the `/config/development.config.php` file is overwriting stuff from the `/config/autoload/zend-expressive.global.php` file. I understand that you need it to make the development mode stuff working. But it is still very confusing that I need to switch between the directory structure to understand the current setting for the caching for example. These settings should either been placed in the `/config/` path or in the `/config/autoload/` path but not in two different paths.

- Currently the routes definitions are placed in the `/config/routes.php` file. The structure contains only the `App` module yet, but think about adding more modules. Users will feel to be forced by the structure to add their routes the `/config/routes.php` file. This is destroying the idea of a modular structure. If I have a `User` module I would expect the routing configuration somewhere in the `/module/User/` path. I know that it is quite hard to build a sensible structure if you want to add the routes programmatically. I love the programmatic addition of the routes but currently this structure needs a little bit of improvement. What about a new method to the `App/ConfigProvider` class which takes the `Zend\Expressive\Application` instance as an parameter? Or what about a `App/RoutesProvider` class?  The `/config/routes.php` should be refactored then to add all these `RoutesProvider` classes or `ConfigProvider::addRoutes()` methods.

- I think the programmatic middleware pipeline is a great step to reduce the complexity of multi-level config files. But I don't like the approach in the `/public/index.php` file to first grab the `Zend\Expressive\Application` instance from the DI container and then requiring the `config/pipeline.php` and `config/routes.php` files which amend the application instance inline. It really looks weird and confusing in the `/public/index.php` file. Why not doing some stuff like `$app = require 'config/application.php';` similar to the configuration of the DI container? We still have the problem how to add the middleware pipeline and the routes programmatically but now it is hidden from such a high level. Inline requiring to add stuff to an existing object feels a bit messy. Another Idea: couldn't we use some custom delegator factories to add stuff like middleware pipeline and routes to the `Zend\Expressive\Application` instance? Then the `/public/index.php` could look like this again:
```
/** @var \Interop\Container\ContainerInterface $container */
$container = require 'config/container.php';

/** @var \Zend\Expressive\Application $app */
$app = $container->get(\Zend\Expressive\Application::class);
$app->run();
```

- Regarding the `composer.json` file. Why not changing the requirements for `Zend\ServiceManager` and `Zend\Stdlib` to its 3.x versions? Is there a sceanario where the 2.x versions are needed? If we force the `Zend\ServiceManager` 3.x versions the configuration can be simplified by removing the `invokables` to teach beginners the new way of  `Zend\ServiceManager` 3.x version.

Ok, this was a longer comment that I thought. Please don't get me wrong. Maybe I am totally on the wrong track. But I think you are doing a great job with `Zend\Expressive` 2.x. Though I think to support the adoption rate for beginners we really need to improve the modular structure of the skeleton. This will lead beginners in the right direction. With the current state most modular `Zend\Expressive` applications will become a mess in a short time. If the skeleton of the new `Zend\Expressive` 2.x will be creating the current state of the application structure this will be the base for a lot of applications. People need some guidance when it comes to structure. 

Feedback and comments are really appreciated.

weierophinney commented:
@xtreamwayz —

> Can we change in test/ExpressiveInstallerTest/ProjectSandboxTrait.php

I just tried this, modifying the code to use the configured `$projectRoot` when including the files, and got a *ton* of test failures and errors. As such, I don't think it's quite safe.

If you can figure out a way to make it work, feel free to send a PR against my branch.

weierophinney commented:
I've updated the code to use the develop branch of Expressive at this time, and removed the `repositories` setting that was pointing to my own development branch.

I had to make the following changes:

- Configure the `DefaultDelegate` and `NotFoundDelegate` in the dependencies.
- Add support for aliases to the Pimple and Aura configuration.
- Decided to add delegator factory support for Aura as well, which requires the installer to copy over a couple additional classes to the `config/` tree.
- Update tests that execute the application to pass the default delegate when invoking the application.

Errors are currently being reported for the `ContainerTest` when used with Aura.Di under PHP 5.6 and when using the lowest dependencies; I'll investigate that in the coming days.

@RalfEggert Thank you for the detailed feedback! I'll start addressing it early next week.

weierophinney commented:
@RalfEggert:

> - Why is the /src/ directory called src at all? It contains classes and templates and the configuration for this module. We expected a /module/ or /modules/ directory instead.

Because it contains your _source_ code. That allows it to be purposed in such as way as to contain both modules and library code, without requiring those be in separate trees of the project.

Since a _module_ in Expressive is simply a tree that is under a common namespace that contains a `ConfigProvider` class, there's nothing else special about it at all. It's just _source code_ for your application.

> - Why is some of the configuration for the App module still placed in /config/autoload/? For example, the dependencies configuration for the two action middleware PingAction and HomePageAction is still placed in /config/autoload/routes.global.php and should be deleted since it is already defined in the App/ConfigProvider. 

Good catch! I'll fix that!

>   It is also confusing that some of the layout config is in the /config/autoload/templates.global.php while the rest is in the App/ConfigProvider class. Please put it in one place only.

I've just reviewed, and I'm not sure I agree with your assessment.

The `ConfigProvider` for the `App` module only defines template paths for templates it provides. The `config/autoload/templates.global.php` file, on the other hand, provides configuration for the chosen template renderer, and _application-level_ configuration — which _can_ include things like a choice of layout (this is only true currently of zend-view; the other template renderers choose the layout within the templates). As such, I think everything is currently in the correct location.

> - If the dependencies configuration for the two action middleware PingAction and HomePageAction is removed from the /config/autoload/routes.global.php file then the RouterInterface::class configuration could be moved to the /config/autoload/dependencies.global.php file and the /config/autoload/routes.global.php file could be deleted altogether, couldn't it? Same for the /config/autoload/templates.global.php when the templates config is moved to the App\ConfigProvider class. Less config file makes it easier for beginners to adopt Zend\Expressive.

I agree with moving the router configuration to the `dependencies.global.php` file. However, as noted above, the `templates.global.php` file contains _application-level_ configuration, which _should not_ be in a module, and I plan to retain it.

> - I find it really confusing that the /config/development.config.php file is overwriting stuff from the /config/autoload/zend-expressive.global.php file. I understand that you need it to make the development mode stuff working. But it is still very confusing that I need to switch between the directory structure to understand the current setting for the caching for example. These settings should either been placed in the /config/ path or in the /config/autoload/ path but not in two different paths.

We have a bit of a chicken-and-egg problem here.

zf-development-mode will do the following when enabling:

- copy `config/development.config.php.dist` to `config/development.config.php`.
- copy `config/autoload/development.local.php.dist` to `config/autoload/development.local.php` _if it exists_.

The problem is that the _first_ file is required; if it does not exist, zf-development-mode will not do anything. So we cannot just remove it.

This is due to its original purpose: enabling development mode in zend-mvc applications, where the development config may also enable additional _modules_, which need to be enabled _first_, as configuration is a two-step process (aggregate all module configuration, then aggregate autoloadable, application-level configuration).

With Expressive, configuration is flat. This means that the two files are roughly equivalent, with one caveat; the configuration from `config/development.config.php` will have precedence over all other settings, as it is merged last; this helps ensure that things like _config caching_ (not general application caching!) and debug mode are enabled.

As such, my thought is this:

- We recommend using `config/autoload/development.local.php.dist` for the majority of development mode settings, which might include things like overrides for caching, logging, databases, etc.
- We ship `config/development.config.php.dist` with only settings for enabling debug mode and disabling config caching, and recommend making no further changes to that file, and note that it exists to allow enabling development mode.

From what I can tell, that would likely address your primary concerns; thoughts?

> - Currently the routes definitions are placed in the /config/routes.php file. The structure contains only the App module yet, but think about adding more modules. Users will feel to be forced by the structure to add their routes the /config/routes.php file. This is destroying the idea of a modular structure. If I have a User module I would expect the routing configuration somewhere in the /module/User/ path. I know that it is quite hard to build a sensible structure if you want to add the routes programmatically. I love the programmatic addition of the routes but currently this structure needs a little bit of improvement. What about a new method to the App/ConfigProvider class which takes the Zend\Expressive\Application instance as an parameter? Or what about a App/RoutesProvider class? The /config/routes.php should be refactored then to add all these RoutesProvider classes or ConfigProvider::addRoutes() methods.

This was something we debated quite a bit before implementing the current solution. Our primary issues with having modules supply routing information come down to usage with third-party modules:

- route strings and options are specific to _routing implementations_, which makes compatibility a problem.
- there is a potential for routing conflicts.
- the routes from a third-party module may not reflect what you actually want to use _in your own application_.

As such, we felt that _not_ having the ability to configure routes was actually a benefit to programmatic applications, as it solves these problems.

The issue, of course, is that you end up with a potential for a lot of copy-and-paste, which can lead to issues when upgrading a module.

Our recommendation thus became: use delegator factories on the `Application` instance.

Part of this pull request is ensuring that delegator factories are supported across each of the supported container implementations. We'd previously supported them on zend-servicemanager (obviously!), as well as Pimple; we now can support them for Aura.Di as well, making this a viable recommendation.

That said, there's a caveat: while internal modules can register them by default, third-party, distributed modules _should not_, for the reasons stated above: conflicts, compatibility, and application specificity. If the routes suggested satisfy your criteria, you could then register the delegator factory without issue; if they do not, you can manage the routes manually within your application.

So, as an example: I could create a distributable module named phly/social-auth that has some default routing. I'd create a delegator factory for `Zend\Expressive\Application` that would register them:

```php
namespace Mwop\SocialAuth;

use Interop\Container\ContainerInterface;
use Zend\Expressive\Application;
use Zend\Expressive\Router\AuraRouter;
use Zend\Expressive\Router\FastRouteRouter;
use Zend\Expressive\Router\ZendRouter;
use Zend\Expressive\Router\RouterInterface;

class SocialAuthApplicationDelegatorFactory
{
    public function __invoke(ContainerInterface $container, $serviceName, callable $callback)
    {
        /** @var Application */
        $application = $callback();

        // Add routes.
        $router = $container->get(RouterInterface::class);

        switch (true) {
            case ($router instanceof AuraRouter):
                $this->injectAuraRoutes($application);
                break;
            case ($router instanceof FastRouteRouter):
                $this->injectFastRouteRoutes($application);
                break;
            case ($router instanceof ZendRouter):
                $this->injectZendRouterRoutes($application);
                break;
            default:
                // Unknown router type; do nothing
                return $application;
        }

        // Return the configured application:
        return $application;
    }
}
```

I would then inform users that they need to enable this delegator factory within their application, _but would not do so by default_.

However, if this were a module _developed directly in my application_, I'd likely:

- skip the bits where I check the router type, and just register routes directly, using the routing syntax for the router I've chosen for my application.
- register the delegator factory within my module's `ConfigProvider`.

Which solves the issues you present: you can keep the routes close to your modules, and have them auto-registered without needing to update the `config/routes.php` file at all.

Make sense?

> - I think the programmatic middleware pipeline is a great step to reduce the complexity of multi-level config files. But I don't like the approach in the /public/index.php file to first grab the Zend\Expressive\Application instance from the DI container and then requiring the config/pipeline.php and config/routes.php files which amend the application instance inline. It really looks weird and confusing in the /public/index.php file. Why not doing some stuff like $app = require 'config/application.php'; similar to the configuration of the DI container? We still have the problem how to add the middleware pipeline and the routes programmatically but now it is hidden from such a high level. Inline requiring to add stuff to an existing object feels a bit messy. Another Idea: couldn't we use some custom delegator factories to add stuff like middleware pipeline and routes to the Zend\Expressive\Application instance? Then the /public/index.php could look like this again:
> 
>   ```php
>   /** @var \Interop\Container\ContainerInterface $container */
>   $container = require 'config/container.php';
>   
>   /** @var \Zend\Expressive\Application $app */
>   $app = $container->get(\Zend\Expressive\Application::class);
>   $app->run();
>   ```

As noted above, you can _also_ use delegator factories with the approach we've provided; as such, _if you want_, you could remove those `require` statements from your `public/index.php` file entirely.

The reason we're using the current approach is because it's far easier to get started by saying "Open the file config/routes.php and add [. . .]" than it is to say "Create a file src/Foo/src/Factory/Routes.php and create the class `Foo\Factory\Routes` with the following contents [. . .]; now update the file config/autoload/dependencies.global.php and add the following [. . .]". On top of being more work, it also requires explaining additional concepts such as delegator factories, modules, etc.

In regards to using something like `require 'config/application.php';`, you end up with the same problem: how is the pipeline created, and how are routes registered? And you either end up with required files, such as we're currently doing, or more advanced concepts such as delegator factories.

One goal for Expressive is to have not only an enterprise-level middleware microframework, but something _approachable_. As such, we feel the `require` approach fulfils the latter goal better.

> - Regarding the composer.json file. Why not changing the requirements for Zend\ServiceManager and Zend\Stdlib to its 3.x versions? Is there a sceanario where the 2.x versions are needed? If we force the Zend\ServiceManager 3.x versions the configuration can be simplified by removing the invokables to teach beginners the new way of Zend\ServiceManager 3.x version.

Good point about updating those dependencies; I'll do so shortly!

That said, we'll still support `invokables` configuration, for one reason: cross-container compability. While zend-servicemanager offers the `InvokableFactory`, which largely removes the need for `invokables`, we do not want to make zend-servicemanager a dependency when other container types are used! (Interestingly, zend-servicemanager, when provided with `invokables` configuration, simply sets up a factory mapping pointing to an `InvokableFactory` instance; if the service name and class name do not match, it also creates an alias.)

Thanks for the great feedback!

weierophinney commented:
> I agree with moving the router configuration to the dependencies.global.php file.

Actually, this is not as easy as it looks.

Currently, the `config/autoload/dependencies.global.php` file is _static_; the information it contains will work for any of the various container, router, or template renderer options. The installer does not touch this file; all it does is _move additional files into that tree_, based on the capabilities selected.

As updating a file _in situ_ is more problematic both in terms of accomplishing as well as _testing_, I'll leave the file in place, minus the middleware dependency registration.

RalfEggert commented:
@weierophinney 

Thanks for your reply. I will pick up some of the points if I have something to add.

> Because it contains your source code. That allows it to be purposed in such as way as to contain both modules and library code, without requiring those be in separate trees of the project.

Ok, that sounds reasonable. However having a path `/src/App/src/Action/` still looks a little unusal. If it only contains source code why do we need to have another `/src` subfolder then? I guess it is needed since we have source code **and** templates within it. Beginners will definitely trip over this and wonder why it is done like this. So I still would prefer `/module/` or something similar. 

> this is only true currently of zend-view; the other template renderers choose the layout within the templates

I think this is the problem with some of the other issues I mentioned. The files in the skeleton are the result of the open concept to support different renderers, routers and DI containers. The file setup makes sense from the point of view of the installer. But people who want to start their project after the installation are really confused about the config data in different places. After we setup the Expressive app last week the first thing we did was tidy up the configuration. From the development point of view it doesn't make sense to have some template configuration in the `/config/autoload/` folder and some in the `App\ConfigProvider` class.

The same applies to the router config you mentioned in the follow up comment.

> We have a bit of a chicken-and-egg problem here.

Ok, I understand the problem but maybe we can have a solution that is less confusing. Could the `/config/development.config.php` be made static and just add another `Zend\ConfigAggregator\PhpFileProvider` instance to add the `config/autoload/{{,*.}development}.php` config files? Then all `development` config stuff can be placed in the `/config/autoload/` folder? 

Anyway we still have another problem, I guess. If we leave it like it is or use this approach, the `development` configuration will overwrite the `local` configuration, won't it? In my MVC projects I always used the levels `global` -> `development` -> `local`, so `local` can overwrite `development` configuration. 

> Our recommendation thus became: use delegator factories on the Application instance.

Yeah, this sounds great. And since I saw that you have already written some docs about it I guess this is a proper solution. Maybe in the files created by the installer itself we could add a link to the docs? 

The same applies for the middleware pipeline.

Thanks again for you thoughts and your reply. I really think we should make it easier for beginners to understand the `Zend\Expressive` way of building applications and make them comfortable from them start. In general, every config file that we don't need is a good config file. If this makes the installer more complicated I really think it is worth it.

xtreamwayz commented:
> Ok, that sounds reasonable. However having a path /src/App/src/Action/ still looks a little unusal.

I don't like it either, but it's copied from zend-mvc. I would prefer something like `src\MyModule\Action` and place all resources like the templates in `src\MyModule\Resources\views` and `src\MyModule\Resources\public`.

As you know, whatever your choice is during installation, you can change it the way you want. You don't need to use the default structure. I don't think there can be a 100% satisfying solution for everyone.

> I think this is the problem with some of the other issues I mentioned. The files in the skeleton are the result of the open concept to support different renderers, routers and DI containers. The file setup makes sense from the point of view of the installer. But people who want to start their project after the installation are really confused about the config data in different places. After we setup the Expressive app last week the first thing we did was tidy up the configuration. From the development point of view it doesn't make sense to have some template configuration in the /config/autoload/ folder and some in the App\ConfigProvider class.
>
> The same applies to the router config you mentioned in the follow up comment.

Those files are the left overs from moving the config to the ConfigProvider class. Those left over dependencies are there for a reason: Simplicity.

There are 3 ways to make the installer work:

1. Collect the entire configuration and dynamically generate all the php files and classes. This would be the nicest solution, however it require to include something like zend-code. Since the installer is executed before anything else and before composer installs any dependencies, zend-code or anything like it can't be used and thus all relevant functionality needs to be copied to the installer and maintained there as well. Unless someone knows an easy to maintain lightweight class to generate PHP code... It would make the installer easier to extend and configure.

2. Make different versions of `config\autoload\dependencies.global.php`. This might be an option as well but it means you need to make a different file for all possible combinations. That's too hard to maintain and it gets even worse if other packages are added in the future.

3. Leave it as it is and keep it simple. For those that don't like it, they can merge the config files and move files around. The installer would still be helpful since it really gets you started within a minute and does all the wiring for you.

Arguably you don't really need the ConfigProvider class if you use the flat structure.

> Maybe in the files created by the installer itself we could add a link to the docs?

Good idea.

weierophinney commented:
@RalfEggert —

>  Beginners will definitely trip over this and wonder why it is done like this. So I still would prefer /module/ or something similar.

When I look at other frameworks, this is what I see:

- Laravel has two concepts: the application source code, which goes under `app/`; and _packages_, which are similar to ZF modules, and go under a `packages/` directory.
- Symfony has _bundles_; these go under the `src/` tree, just like any other source code.
- Slim puts everything under `src/` including non-OOP code, but does not have a concept of modules or packages.
- ZF has a `module/` directory, because all source code goes in modules.

I'd argue that if a `src/` tree that combines bundles and regular source code is okay for Symfony, we can do the same here. If end-users want to structure differently, _of course they can_; we do not dictate where code goes, we simply provide a default location so you have something to work with immediately.

> But people who want to start their project after the installation are really confused about the config data in different places. After we setup the Expressive app last week the first thing we did was tidy up the configuration.

Who is "really confused"? I'm not seeing a lot of chatter on the issue tracker about confusion, to be honest, so if _you_ are seeing this, I'd love to see some links so I can investigate further.

I think the different files made more sense when we were doing _more_ in them. The `middleware-pipeline.global.php` had both dependencies for the various middleware in the pipeline, as well as configuration for it; the `routes.global.php` file had dependency configuration for routed middleware, as well as configuration for the routes itself.

I feel `templates.global.php` still works in this regard; it has dependency configuration for the renderer being used, as well as configuration specific to that template renderer. (More on this below.)

But I think having multiple files fills another niche: it helps demonstrate that you don't have to put all configuration in a single file, but can separate it by semantic needs. This is great for maintainability and understanding how various pieces interact — and with configuration caching enabled by default in production, there's no expensive overhead.

> From the development point of view it doesn't make sense to have some template configuration in the /config/autoload/ folder and some in the App\ConfigProvider class.

Again, _there is no template configuration_ in the `config/autoload/templates.global.php` file. It contains _only_ dependency configuration, and configuration options for the specified template renderer: what extension to look for, information on caching, assets, extensions, and, yes, in the case of zend-view, the default layout to use (as it's the only renderer that doesn't specify the layout directly in the template).

> (something about autoloading development files using the aggregator)

You can, of course, do that _in your own project_. However, the idea around zf-development-mode is to allow you to quickly and easily switch in and out of a set of specific development configurations, specified in _two specific files_. If you have a `PhpFileProvider` aggregating anything with a `.development.php` extension _always_, you then have to manually manage those files (i.e., copying them from `.dist` files to non-`.dist` files, etc.). It's out-of-scope.

> the development configuration will overwrite the local configuration, won't it?

Depends!

If you do not alter `config/development.config.php.dist`, then that only overrides the configuration caching and debug settings. Any other overrides will be in `config/autoload/development.config.php.dist` — and you can have another `.local.php` file override _those_ settings by having it named such that it will glob later. As an example, `zzz.local.php` will autoload later.

> Maybe in the files created by the installer itself we could add a link to the docs?

Sure!

---

@xtreamwayz —

> I would prefer something like src\MyModule\Action and place all resources like the templates in src\MyModule\Resources\views and src\MyModule\Resources\public.

I've never been a fan of mixing non-PHP code in the same tree as the PHP code itself, which is why ZF modules have a `src/` directory, and then peer directories for configuration, views, etc.

That said, I see that Symfony bundles do exactly this, so I have no argument there. Would you care to send a PR against my branch to enable this?

RalfEggert commented:
@xtreamwayz 

> I don't like it either, but it's copied from zend-mvc.

Well, I think it is not. Zend\Mvc uses a `/module/Application/` path with `src/`, `view/` and `config/` in it. It doesn't use `/src/Application/`.

https://github.com/zendframework/ZendSkeletonApplication/tree/master/module/Application

> 3. Leave it as it is and keep it simple. For those that don't like it, they can merge the config files and move files around. The installer would still be helpful since it really gets you started within a minute and does all the wiring for you.

I understand your points. Ok, then we definitely need a recipe what users can do to simplify the config files after installation. I am willing to write this recipe when everything is settled and I find the time.

RalfEggert commented:
@weierophinney 

> When I look at other frameworks, this is what I see

The reason why I like the Zend Framework (1 to 3) since the beginning is its clear structure. With all the other frameworks you mentioned I always had some problems to find my way around. Maybe I am getting old and becoming a creature of habits. I never understood the separation of `app` and `src` in Symfony both having source code and templates in it. And maybe I am just used to the `Zend\Mvc` way of structuring its models with `Zend\ModuleManager`. If a `/src` folder with both source code, config and templates is the way to go, I will start to live with it. At least I know how to change the structure. ;-)

> Who is "really confused"? I'm not seeing a lot of chatter on the issue tracker about confusion, to be honest, so if you are seeing this, I'd love to see some links so I can investigate further.

Well, I moderated two ZF3 and `Zend\Expressive` workshops in the last weeks. In both of these workshops it took me a long time to explain the structure of the config files as it is. In both of these workshops the first things we did was some reorganizing. So there will be no links I can present. It is just some feedback from experience I personally made.

And don't get me wrong. I am not against multiple files for different aspects of configurations.

xtreamwayz commented:
@weierophinney 

> That said, I see that Symfony bundles do exactly this, so I have no argument there. Would you care to send a PR against my branch to enable this?

It doesn't make sense to do that. If you change the module structure `/src/App/src/Action` to `/src/App/Action`, you would have the flat structure.

That's why I proposed somewhere to use `/src/Action` as the flat structure.

@RalfEggert 

> I understand your points. Ok, then we definitely need a recipe what users can do to simplify the config files after installation. I am willing to write this recipe when everything is settled and I find the time.

I don't think you need to write a recipe for this. It's more a personal taste on how to fine tune your application. I've seen several structures already and one has all local and global files grouped in different directories. You could even argue to split the dependencies.global.php file in smaller files to group dependencies. It might be easier in larger applications.

> Well, I think it is not. Zend\Mvc uses a /module/Application/ path with src/, view/ and config/ in it. It doesn't use /src/Application/.

Actually, I thought you tripped over the two times `src` in the module structure. That's why I came up with a Symfony like approach to get rid of the second one. I still think that source code belongs in `/src` rather than in `/module`. But again, this my personal preference and this skeleton can't please everyone.

RalfEggert commented:
@xtreamwayz 

> But again, this my personal preference and this skeleton can't please everyone.

Yep, I think that is a good final summary. I will try to start to refrain from pushing my personal preferences from the skeleton. 

I really think, `Zend\Expressive` is the best shit since sliced bread. I hope I bought in some points to think about in this discussion though.

xorock commented:
I think whole structure should be the same as in ZF3. Beginners have an opinion and simply adapts to the proposed structure. For more advanced programmers it just simplifies using modules between these 2 frameworks. I have just finished quick project on Expressive and it was very convenient to copy finished modules from ZF3 with some quick adaptation. My structure is:
```
/module/namespace/config[/module.config.php|someotherconfig.php]
|-.../src[/Action|Entity|Form|Hydrator|...]
|-.../view
```
Template path inside module config is set to ./view within namespace.

More confusing, after reading 1.1 changelog, was - why do I have to use programatic pipeline when everything was set in config files earlier.

weierophinney commented:
@xorock —

> More confusing, after reading 1.1 changelog, was - why do I have to use programatic pipeline when everything was set in config files earlier.

Because we have found through teaching and support that the developers find the programmatic pipeline and routing to be easier to understand and easier to create than the configuration files. We originally used configuration because of our background with zend-mvc, but there was a disconnect when applying it to middleware applications.

> Docs for Expressive are the weakest point, and valuable insights are mostly here on github. If You do not track issues then docs are not a good place to start.

If you feel the documentation should be improved, please submit issues or pull requests against the zend-expressive repository with your suggestions.

danizord commented on src/ExpressiveInstaller/Resources/src/ExpressiveAuraConfig.php:
```diff
@@ -0,0 +1,163 @@
+ * - Invokable classes
+ * - Services (known instances)
+ */
+class ExpressiveAuraConfig implements ContainerConfigInterface
```
@weierophinney I see that you created a layer that makes all supported containers compatible with SM-like config. What about moving this layer to a "expressive-container-interop" package? I think that this is usable as well by other frameworks :)

weierophinney replied:
That's what [service-provider](https://github.com/container-interop/service-provider) is tackling. :)

The idea then will be that container providers would consume the service-provider interface in order to wire services to the container. The way the interface is written, it can be either a straight factory, or act as a delegator, which solves the majority of problems (invokables and aliases may be accomplished using a combination of these).

My plan is to bring some our ideas and findings to that project/working group, instead of creating a project of our own.

weierophinney commented:
> I hope I bought in some points to think about in this discussion though.

You absolutely did! Thanks for collaborating!
